# Задание 1. Базовое изучение

In [107]:
import pandas as pd
import numpy as np

In [139]:
df2 = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv',
                header=None, names=["1. oper", "2. vozrast", "3. #bolnici", "temp_rekt", "pulce", 
                                    "chast_dihaniya", "7. temp kon", "8. perif imp", "9. sliz o", 
                                    "10. time zap kap", "11. bol'", "12. peristaltica", "13. vzdut jivot", 
                                    "14. naz gaz", "15. naz ref", "16. naz ref PH", "17. kal", "18. bruh pol", 
                                    "19. V yachey", "20. belok", "21. abdomin", "22. belok brusch", "23. rezult", 
                                    "24. hirurg povr", "25.", "26.", "27", "cp_data"], na_values = True
                 )
df2.head()

1. oper  2. vozrast  3. #bolnici temp_rekt pulce chast_dihaniya 7. temp kon  \
0       2         NaN       530101     38.50    66             28           3   
1     NaN         NaN       534817      39.2    88             20           ?   
2       2         NaN       530334     38.30    40             24         NaN   
3     NaN         9.0      5290409     39.10   164             84           4   
4       2         NaN       530255     37.30   104             35           ?   

  8. perif imp 9. sliz o 10. time zap kap  ... 19. V yachey 20. belok  \
0            3         ?                2  ...        45.00      8.40   
1            ?         4              NaN  ...           50        85   
2          NaN         3              NaN  ...        33.00      6.70   
3          NaN         6                2  ...        48.00      7.20   
4            ?         6                2  ...        74.00      7.40   

  21. abdomin 22. belok brusch 23. rezult 24. hirurg povr    25. 26. 27  \
0           ?                ?          2             2.0  11300   0  0   
1           2                2          3             2.0   2208   0  0   
2           ?                ?        NaN             2.0      0   0  0   
3           3             5.30          2             NaN   2208   0  0   
4           ?                ?          2             2.0   4300   0  0   

  cp_data  
0     2.0  
1     2.0  
2     NaN  
3     NaN  
4     2.0  

[5 rows x 28 columns]

In [147]:
df = df2.drop(["3. #bolnici", "7. temp kon",  "9. sliz o", "10. time zap kap", "11. bol'", "12. peristaltica", 
         "13. vzdut jivot", "14. naz gaz", "15. naz ref", "16. naz ref PH", "17. kal", "18. bruh pol", 
         "19. V yachey", "20. belok", "21. abdomin", "22. belok brusch", "25.", "26.", "27", "cp_data"]
                 , axis = 1)

In [148]:
df[["temp_rekt", "pulce", "chast_dihaniya"]] = df[["temp_rekt", "pulce", "chast_dihaniya"]].apply(pd.to_numeric, errors='coerce')
df

1. oper  2. vozrast  temp_rekt  pulce  chast_dihaniya 8. perif imp  \
0         2         NaN       38.5   66.0            28.0            3   
1       NaN         NaN       39.2   88.0            20.0            ?   
2         2         NaN       38.3   40.0            24.0          NaN   
3       NaN         9.0       39.1  164.0            84.0          NaN   
4         2         NaN       37.3  104.0            35.0            ?   
..      ...         ...        ...    ...             ...          ...   
295     NaN         NaN        NaN  120.0            70.0            ?   
296       2         NaN       37.2   72.0            24.0            2   
297     NaN         NaN       37.5   72.0            30.0            3   
298     NaN         NaN       36.5  100.0            24.0            3   
299     NaN         NaN       37.2   40.0            20.0            ?   

    23. rezult  24. hirurg povr  
0            2              2.0  
1            3              2.0  
2          NaN              2.0  
3            2              NaN  
4            2              2.0  
..         ...              ...  
295          3              2.0  
296          3              NaN  
297          2              NaN  
298        NaN              NaN  
299          3              2.0  

[300 rows x 8 columns]

## Базовые метрики

In [134]:
print('Базовые метрики по тепературе: ')
print('Температура, max: ', max(df['temp_rekt']))
print('Температура, min: ', min(df['temp_rekt']))
print('Среднеарифметическое значение: ', df.temp_rekt.mean())
print('Медиана: ', df.temp_rekt.median())
print('Значение моды: ', df.temp_rekt.mode()[0])
print('СКО: ', df.temp_rekt.std())
print('Дисперсия: ', df.temp_rekt.var())
print('______________________________________________________________')      
print('Базовые метрики по пульсу:')
print('Пульс, max: ', np.max(df['pulce']))
print('Пульс, min: ', np.min(df['pulce']))
print('Среднеарифметическое значение: ', df.pulce.mean())
print('Медиана: ', df.pulce.median())
print('Значение моды: ', df.pulce.mode()[0])
print('СКО: ', df.pulce.std())
print('Дисперсия: ', df.pulce.var())
print('______________________________________________________________')       
print('Базовые метрики по частоте дыхания')
print('Частота дыхания, max', df['chast_dihaniya'].max())
print('Частота дыхания, min', df['chast_dihaniya'].min())
print('Среднеарифметическое значение: ', df.chast_dihaniya.mean())
print('Медиана: ', df.chast_dihaniya.median())
print('Значение моды: ', df.chast_dihaniya.mode()[0])
print('СКО: ', df.chast_dihaniya.std())
print('Дисперсия: ', df.chast_dihaniya.var())      

Базовые метрики по тепературе: 
Температура, max:  40.8
Температура, min:  35.4
Среднеарифметическое значение:  38.16791666666669
Медиана:  38.2
Значение моды:  38.0
СКО:  0.7322886641121578
Дисперсия:  0.5362466875871686
______________________________________________________________
Базовые метрики по пульсу:
Пульс, max:  184.0
Пульс, min:  30.0
Среднеарифметическое значение:  71.91304347826087
Медиана:  64.0
Значение моды:  48.0
СКО:  28.630556660735003
Дисперсия:  819.7087747035575
______________________________________________________________
Базовые метрики по частоте дыхания
Частота дыхания, max 96.0
Частота дыхания, min 8.0
Среднеарифметическое значение:  30.417355371900825
Медиана:  24.5
Значение моды:  20.0
СКО:  17.642231385134664
Дисперсия:  311.24832824663054


#### Согласно базовым метрикам имеется существенное различие между минимальным и максимальным значениями пульса и частоты дыхакния, и как следствие высокие показатели СКО и дисперсии. 
#### Минимальное и максимальное значение температуры не имеет столь существенного различия, что влияет на показатели СКО и дисперсии. Отсутствие существенного различие в данных обосновывается физиологией животных, т.к. слишком высокие/низкие показатели температуры приводят к смерти животного и соответственно нет необходимости/возможности измерения прочих параметров. В случае наличия большей выборки по температуре, то подобные минимальное/максимальное значения можно принять как пограничные (живое/мертвое животное).

# Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [151]:
df3 = df.drop(["1. oper", "2. vozrast", "8. perif imp", "23. rezult", "24. hirurg povr"]
                 , axis = 1)
print(df['temp_rekt'].mean()) # среднее
print(df[(df['temp_rekt'] > 35.7) & (df['temp_rekt'] < 39.5)]['temp_rekt'].mean())
print(df['temp_rekt'].median()) #медиана
print(df[(df['temp_rekt'] > 35.7) & (df['temp_rekt'] < 39.5)]['temp_rekt'].median())
print(df['temp_rekt'].round().mode()[0]) #мода
print(df[(df['temp_rekt'] > 35.7) & (df['temp_rekt'] < 39.5)]['temp_rekt'].round().mode()[0])

38.16791666666669
38.09780701754388
38.2
38.1
38.0
38.0


In [153]:
q1 = df3['temp_rekt'].quantile(0.25)
q3 = df3['temp_rekt'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df3[df3['temp_rekt'].between(lower_bound, upper_bound, inclusive=True)].sort_values('pulce')
remove_outliers

temp_rekt  pulce  chast_dihaniya
130       37.8   30.0            12.0
232       38.5   30.0            18.0
276       38.0   36.0            16.0
203       37.2   36.0             9.0
76        37.6   38.0            20.0
..         ...    ...             ...
117       39.5    NaN             NaN
126       38.7    NaN             NaN
151       39.3    NaN             NaN
182       38.0    NaN            24.0
204       39.2    NaN            23.0

[226 rows x 3 columns]

In [163]:
pd.concat([df3, remove_outliers]).drop_duplicates(keep=False)#.count()

temp_rekt  pulce  chast_dihaniya
7          NaN   60.0             NaN
8          NaN   80.0            36.0
16         NaN  128.0            36.0
20        39.9   72.0            60.0
35         NaN  104.0            24.0
40         NaN   88.0             NaN
44        35.4  140.0            24.0
57         NaN   64.0            36.0
58         NaN    NaN            20.0
65         NaN   60.0            36.0
68         NaN   52.0            16.0
75        39.7  100.0             NaN
80        36.4   98.0            35.0
86         NaN   88.0            40.0
87         NaN   40.0            16.0
96         NaN   84.0            36.0
99        39.6  108.0            51.0
101        NaN   60.0            20.0
105        NaN   70.0            36.0
112        NaN   40.0             NaN
115        NaN    NaN            40.0
118       36.5   78.0            30.0
141       36.0   42.0            30.0
145        NaN   80.0            32.0
159        NaN    NaN            12.0
167        NaN   44.0            24.0
169        NaN   65.0            24.0
173        NaN  104.0             NaN
177        NaN   68.0            14.0
178        NaN   60.0            30.0
191        NaN  100.0            44.0
193        NaN   60.0            48.0
219        NaN   52.0            28.0
222        NaN  112.0            13.0
238       36.1   88.0             NaN
242        NaN   36.0            32.0
251       36.6   42.0            18.0
253        NaN  112.0            24.0
254        NaN   80.0             NaN
259       40.8   72.0            42.0
265        NaN   56.0            68.0
274        NaN   76.0             NaN
281       40.0   78.0             NaN
293        NaN   78.0            24.0
295        NaN  120.0            70.0
298       36.5  100.0            24.0

#### Обнаружено 12 выбросов по температуре лошадей(менее 36.6 и более 39.6 ), при этом нормальная температура лошадей находится в диапазоне 37,5 -38,5 градусов.
#### В зависимости от цели данные выбросы можно исключить из выборки (при поиске иных более существенных факторов, влияющих на общее состояние лошади), не исключать (в целях понимания многофакторной взаимосвязи показателей), либо работать исключительно с данным выбросами паралельно анализируюя прочие параметры с отколонениями.

In [162]:
print(remove_outliers['temp_rekt'].mean())
print(remove_outliers['temp_rekt'].median())
print(df3['temp_rekt'].mean() - remove_outliers['temp_rekt'].mean())
print(df3['temp_rekt'].median() - remove_outliers['temp_rekt'].median())

38.169026548672576
38.2
-0.0011098820058847991
0.0


#### Влияние данных выбросов на среднее значение минимально, на медиану равно нулю.

In [193]:
print(df['pulce'].mean()) # среднее
print(df[(df['pulce'] > 60) & (df['pulce'] < 150)]['pulce'].mean())
print(df['pulce'].median()) #медиана
print(df[(df['pulce'] > 60) & (df['pulce'] < 150)]['pulce'].median())
print(df['pulce'].round().mode()[0]) #мода
print(df[(df['pulce'] > 60) & (df['pulce'] < 150)]['pulce'].round().mode()[0])

q1 = df3['pulce'].quantile(0.25)
q3 = df3['pulce'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers1 = df3[df3['pulce'].between(lower_bound, upper_bound, inclusive=True)].sort_values('temp_rekt')
remove_outliers1

71.91304347826087
91.14598540145985
64.0
88.0
48.0
88.0


temp_rekt  pulce  chast_dihaniya
44        35.4  140.0            24.0
141       36.0   42.0            30.0
238       36.1   88.0             NaN
80        36.4   98.0            35.0
298       36.5  100.0            24.0
..         ...    ...             ...
272        NaN  100.0            30.0
274        NaN   76.0             NaN
282        NaN   70.0            16.0
293        NaN   78.0            24.0
295        NaN  120.0            70.0

[271 rows x 3 columns]

In [194]:
pd.concat([df3, remove_outliers1]).drop_duplicates(keep=False)#.count()

temp_rekt  pulce  chast_dihaniya
3         39.1  164.0            84.0
41        39.0  150.0            72.0
55        38.6  160.0            20.0
58         NaN    NaN            20.0
115        NaN    NaN            40.0
117       39.5    NaN             NaN
126       38.7    NaN             NaN
151       39.3    NaN             NaN
159        NaN    NaN            12.0
204       39.2    NaN            23.0
255       38.8  184.0            84.0
275       38.8  150.0            50.0

#### Обнаружено 5 выбросов по пульсу лошадей(более 150 ), при этом нормальный пульс в состоянии покоя лошадей находится в диапазоне от 25 до 40. Наличие большого количества лошадей с пульсом сверх нормы свидетельствет о болезни лошадей либо измерением пульса в процессе бега.
#### Исключение выбросов не целесообразно, т.к. возможно свокупность факторов(пульс, температура и пр. данные) может стать более показательной при определении резулбьатов исследования патологий животных.

In [196]:
print(remove_outliers['pulce'].mean())
print(remove_outliers['pulce'].median())
print(df3['pulce'].mean() - remove_outliers1['pulce'].mean())
print(df3['pulce'].median() - remove_outliers1['pulce'].median())

70.25830258302582
64.0
1.6547408952350509
0.0


#### Влияние данных выбросов на среднее значение незначительно, на медиану равно нулю.

In [176]:
print(df['chast_dihaniya'].mean()) # среднее
print(df[(df['chast_dihaniya'] > 60) & (df['chast_dihaniya'] < 150)]['chast_dihaniya'].mean())
print(df['pulce'].median()) #медиана
print(df[(df['chast_dihaniya'] > 60) & (df['chast_dihaniya'] < 150)]['chast_dihaniya'].median())
print(df['chast_dihaniya'].round().mode()[0]) #мода
print(df[(df['chast_dihaniya'] > 60) & (df['chast_dihaniya'] < 150)]['chast_dihaniya'].round().mode()[0])

q1 = df3['chast_dihaniya'].quantile(0.25)
q3 = df3['chast_dihaniya'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers2 = df3[df3['chast_dihaniya'].between(lower_bound, upper_bound, inclusive=True)].sort_values('pulce')
remove_outliers2

30.417355371900825
79.41176470588235
64.0
80.0
20.0
68.0


temp_rekt  pulce  chast_dihaniya
232       38.5   30.0            18.0
130       37.8   30.0            12.0
276       38.0   36.0            16.0
242        NaN   36.0            32.0
203       37.2   36.0             9.0
..         ...    ...             ...
83        38.0    NaN            24.0
115        NaN    NaN            40.0
159        NaN    NaN            12.0
182       38.0    NaN            24.0
204       39.2    NaN            23.0

[225 rows x 3 columns]

In [186]:
pd.concat([df3, remove_outliers3]).drop_duplicates(keep=False).sort_values('chast_dihaniya')#.count()

temp_rekt  pulce  chast_dihaniya
120       39.4   54.0            66.0
103       38.0  140.0            68.0
125       38.0   42.0            68.0
265        NaN   56.0            68.0
229       38.5  120.0            70.0
295        NaN  120.0            70.0
41        39.0  150.0            72.0
82        38.1  100.0            80.0
84        37.8   60.0            80.0
208       37.8   88.0            80.0
3         39.1  164.0            84.0
255       38.8  184.0            84.0
244       38.2  124.0            88.0
39        39.2  146.0            96.0
106       38.3   52.0            96.0
7          NaN   60.0             NaN
9         38.3   90.0             NaN
29        37.7   48.0             NaN
32        37.2   60.0             NaN
37        37.8   72.0             NaN
38        38.6   52.0             NaN
40         NaN   88.0             NaN
50        38.6   46.0             NaN
66        37.9   72.0             NaN
72        37.7   56.0             NaN
75        39.7  100.0             NaN
81        37.3   40.0             NaN
112        NaN   40.0             NaN
117       39.5    NaN             NaN
126       38.7    NaN             NaN
129       38.4   49.0             NaN
137       38.0   48.0             NaN
138       38.0   56.0             NaN
148       38.3  132.0             NaN
151       39.3    NaN             NaN
164       38.7   76.0             NaN
173        NaN  104.0             NaN
201       38.5   86.0             NaN
238       36.1   88.0             NaN
254        NaN   80.0             NaN
256       37.5   72.0             NaN
271       37.7   80.0             NaN
274        NaN   76.0             NaN
281       40.0   78.0             NaN
284       38.5   54.0             NaN

###### Обнаружено 15 выбросов по частоте дыхания(более 66 ), при этом нормальная частота дыхания от 8 до 12 вздохов, у новорожденных  от 60 до 80. В среднем в выборке частота составляет 30 вздохов.
#### Исключение выбросов не целесообразно, необходимо разделение выборки на старых и молодых лошадей для получения достоверного результата.

In [188]:
print(remove_outliers['chast_dihaniya'].mean())
print(remove_outliers['chast_dihaniya'].median())
print(df3['chast_dihaniya'].mean() - remove_outliers3['chast_dihaniya'].mean())
print(df3['chast_dihaniya'].median() - remove_outliers3['chast_dihaniya'].median())

29.904761904761905
25.0
3.701799816345268
0.5


#### Влияние данных выбросов на среднее значение несущественно, на медиану незначительно.

# Задание 3. Работа с пропусками

In [201]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   1. oper          120 non-null    object 
 1   2. vozrast       24 non-null     float64
 2   temp_rekt        240 non-null    float64
 3   pulce            276 non-null    float64
 4   chast_dihaniya   242 non-null    float64
 5   8. perif imp     185 non-null    object 
 6   23. rezult       122 non-null    object 
 7   24. hirurg povr  109 non-null    float64
dtypes: float64(5), object(3)
memory usage: 18.9+ KB


#### В представленной выборке наибольшее количество пропусков находится в столбце "2. vozrast", а также столбце 24. hirurg povr. Необходимо удалить данные столбцы по причине недостаточности данных. В целом выборка имеет слишком большое количество пропусков для принятия взвешенного решения о здоровье лошадей на основании имеющихся данных.

In [234]:
df4 = df.drop(['2. vozrast','24. hirurg povr'], axis=1)#.info()

In [235]:
df4.head()

1. oper  temp_rekt  pulce  chast_dihaniya 8. perif imp 23. rezult
0       2       38.5   66.0            28.0            3          2
1     NaN       39.2   88.0            20.0            ?          3
2       2       38.3   40.0            24.0          NaN        NaN
3     NaN       39.1  164.0            84.0          NaN          2
4       2       37.3  104.0            35.0            ?          2

## Датафрейм без пропусков

In [240]:
df4.dropna()#.info() 

1. oper  temp_rekt  pulce  chast_dihaniya 8. perif imp 23. rezult
0         2       38.5   66.0            28.0            3          2
4         2       37.3  104.0            35.0            ?          2
30        2       37.7   96.0            30.0            3          2
31        2       37.2  108.0            12.0            3          3
36        2       38.3  112.0            16.0            3          3
49        2       37.2   84.0            48.0            3          2
54        2       40.3  114.0            36.0            3          3
63        2       38.2  130.0            16.0            3          3
70        2       37.0   60.0            20.0            ?          3
91        2       40.3  114.0            36.0            3          2
141       2       36.0   42.0            30.0            ?          2
171       2       39.0   86.0            16.0            3          2
197       2       38.5   96.0            36.0            3          2
231       2       39.5   92.0            28.0            3          2
251       2       36.6   42.0            18.0            3          2
296       2       37.2   72.0            24.0            2          3

In [239]:
print('Среднеарифметическое значение: ', df.temp_rekt.mean())
print('Медиана: ', df.temp_rekt.median())
print('Значение моды: ', df.temp_rekt.mode()[0])
print('СКО: ', df.temp_rekt.std())
print('Дисперсия: ', df.temp_rekt.var())
print('______________________________________________________________') 

print(df4.dropna()['temp_rekt'].mean())
print(df4.dropna()['temp_rekt'].median())
print(df4.dropna()['temp_rekt'].mode()[0])

print(df4.dropna()['temp_rekt'].std())
print(df4.dropna()['temp_rekt'].var())

Среднеарифметическое значение:  38.16791666666669
Медиана:  38.2
Значение моды:  38.0
СКО:  0.7322886641121578
Дисперсия:  0.5362466875871686
______________________________________________________________
38.050000000000004
37.95
37.2
1.2617448236469992
1.5919999999999974


In [ ]:
df2 = pd.DataFrame(columns=["1. oper", "2. vozrast", "3. #bolnici", "4. temp rekt", "5. pulce", "6. chast dihaniya", "7. temp kon", "8. perif imp", "9. sliz o", "10. time zap kap", "11. bol'", "12. peristaltica", "13. vzdut jivot", "14. naz gaz", "15. naz ref", "16. naz ref PH", "17. kal", "18. bruh pol", "19. V yachey", "20. belok", "21. abdomin", "22. belok brusch", "23. rezult", "24. hirurg povr", "25.", "26.", "27", "cp_data"])
df2.head()